test/TwabLib.t.sol

```diff
@@ -714,25 +687,45 @@ contract TwabLibTest is BaseTest {
 
   // ================== getPreviousOrAtObservation ==================
 
+  function testGetPreviousOrAtObservation_single_before() public {
+    uint32 t0 = PERIOD_OFFSET;
+    twabLibMock.increaseBalances(0, 1e18);
+    assertEq(twabLibMock.getPreviousOrAtObservation(t0 - 1).timestamp, t0-1);
+  }
+
+  function testGetPreviousOrAtObservation_single_at() public {
+    uint32 t0 = PERIOD_OFFSET;
+    twabLibMock.increaseBalances(0, 1e18);
+    ObservationLib.Observation memory obs = twabLibMock.getPreviousOrAtObservation(t0);
+    assertEq(obs.timestamp, t0);
+  }
+
+  function testGetPreviousOrAtObservation_single_after() public {
+    uint32 t0 = PERIOD_OFFSET;
+    twabLibMock.increaseBalances(0, 1e18);
+    ObservationLib.Observation memory obs = twabLibMock.getPreviousOrAtObservation(t0+1);
+    assertEq(obs.timestamp, t0);
+  }
+
   function testGetPreviousOrAtObservation() public {
-    uint32 t0 = PERIOD_OFFSET + 10 seconds;
+    uint32 t0 = PERIOD_OFFSET;
     uint32 t1 = PERIOD_OFFSET + PERIOD_LENGTH;
     uint32 t2 = PERIOD_OFFSET + (PERIOD_LENGTH * 2);
     uint32 t3 = PERIOD_OFFSET + (PERIOD_LENGTH * 3);
     ObservationLib.Observation memory prevOrAtObservation;
 
     vm.warp(t0);
     twabLibMock.increaseBalances(1, 1);
+    uint16 index;
 
     // Get observation at timestamp before first observation
-    prevOrAtObservation = twabLibMock.getPreviousOrAtObservation(t0 - 1 seconds);
-    assertEq(prevOrAtObservation.timestamp, PERIOD_OFFSET);
+    assertEq(twabLibMock.getPreviousOrAtObservation(t0 - 1 seconds).timestamp, t0-1);
```

Comment:
Shouldn't it return `0` since there is no observation before this time?

Reply:
It actually makes up an observation at the given timestamp when there is no history